core/dbt/rpc/task_manager.py

```diff
@@ -0,0 +1,196 @@
+        self.config = config
+        self.tasks = {}
+        self.completed = {}
+        self._rpc_task_map = {}
+        self._last_compile = LastCompile(status=ManifestStatus.Init)
+        self._lock = multiprocessing.Lock()
+
+    def add_request(self, request_handler):
+        self.tasks[request_handler.task_id] = request_handler
+
+    def reserve_handler(self, task):
+        self._rpc_task_map[task.METHOD_NAME] = None
+
+    def add_task_handler(self, task, manifest):
+        self._rpc_task_map[task.METHOD_NAME] = task(
+            self.args, self.config, manifest
+        )
+
+    def rpc_task(self, method_name):
+        with self._lock:
+            return self._rpc_task_map[method_name]
+
+    def ready(self):
+        with self._lock:
+            return self._last_compile.status == ManifestStatus.Ready
+
+    def set_compiling(self):
+        assert self._last_compile.status != ManifestStatus.Compiling, \
+            f'invalid state {self._last_compile.status}'
+        with self._lock:
+            self._last_compile = LastCompile(status=ManifestStatus.Compiling)
+
+    def set_compile_exception(self, exc, logs=List[Dict[str, Any]]):
+        assert self._last_compile.status == ManifestStatus.Compiling, \
+            f'invalid state {self._last_compile.status}'
+        self._last_compile = LastCompile(
+            error={'message': str(exc)},
+            status=ManifestStatus.Error,
+            logs=logs
+        )
+
+    def set_ready(self, logs=List[Dict[str, Any]]):
+        assert self._last_compile.status == ManifestStatus.Compiling, \
+            f'invalid state {self._last_compile.status}'
+        self._last_compile = LastCompile(
+            status=ManifestStatus.Ready,
+            logs=logs
+        )
+
+    def process_status(self):
+        with self._lock:
+            last_compile = self._last_compile
+
+        status = last_compile.to_dict()
+        status['pid'] = os.getpid()
+        return status
+
+    def process_listing(self, active=True, completed=False):
+        included_tasks = {}
+        with self._lock:
+            if completed:
+                included_tasks.update(self.completed)
+            if active:
+                included_tasks.update(self.tasks)
+
+        table = []
+        now = time.time()
+        for task_handler in included_tasks.values():
+            start = task_handler.started
+            if start is not None:
+                elapsed = now - start
+
+            table.append(TaskRow(
+                str(task_handler.task_id), task_handler.request_id,
+                task_handler.request_source, task_handler.method,
+                task_handler.state, start, elapsed, task_handler.timeout
+            ))
+        table.sort(key=lambda r: (r.state, r.start))
+        result = {
+            'rows': [dict(r._asdict()) for r in table],
+        }
+        return result
+
+    def process_kill(self, task_id):
+        # TODO: this result design is terrible
```

Comment:
is this todo for this pr, or a future one?

Reply:
This is an old comment that I copied over, but it's still true - so a future PR.